sycl/plugins/unified_runtime/ur/adapters/hip/context.cpp

```diff
@@ -0,0 +1,174 @@
+
+#include "context.hpp"
+
+/// Create a UR HIP context.
+///
+/// By default creates a scoped context and keeps the last active HIP context
+/// on top of the HIP context stack.
+///
+UR_APIEXPORT ur_result_t UR_APICALL
+urContextCreate(uint32_t DeviceCount, const ur_device_handle_t *phDevices,
+                const ur_context_properties_t *pProperties,
+                ur_context_handle_t *phContext) {
+  std::ignore = DeviceCount;
+  std::ignore = pProperties;
+
+  assert(DeviceCount == 1);
+  ur_result_t RetErr = UR_RESULT_SUCCESS;
+
+  std::unique_ptr<ur_context_handle_t_> ContextPtr{nullptr};
+  try {
+    hipCtx_t Current = nullptr;
+
+    // Create a scoped context.
+    hipCtx_t NewContext;
+    UR_CHECK_ERROR(hipCtxGetCurrent(&Current));
+    RetErr = UR_CHECK_ERROR(
+        hipCtxCreate(&NewContext, hipDeviceMapHost, phDevices[0]->get()));
+    ContextPtr = std::unique_ptr<ur_context_handle_t_>(new ur_context_handle_t_{
+        ur_context_handle_t_::kind::UserDefined, NewContext, *phDevices});
+
+    static std::once_flag InitFlag;
+    std::call_once(
+        InitFlag,
+        [](ur_result_t &) {
+          // Use default stream to record base event counter
+          UR_CHECK_ERROR(hipEventCreateWithFlags(&ur_platform_handle_t_::EvBase,
+                                                 hipEventDefault));
+          UR_CHECK_ERROR(hipEventRecord(ur_platform_handle_t_::EvBase, 0));
+        },
+        RetErr);
+
+    // For non-primary scoped contexts keep the last active on top of the stack
+    // as `cuCtxCreate` replaces it implicitly otherwise.
+    // Primary contexts are kept on top of the stack, so the previous context
+    // is not queried and therefore not recovered.
+    if (Current != nullptr) {
+      UR_CHECK_ERROR(hipCtxSetCurrent(Current));
+    }
+
+    *phContext = ContextPtr.release();
+  } catch (ur_result_t Err) {
+    RetErr = Err;
+  } catch (...) {
+    RetErr = UR_RESULT_ERROR_OUT_OF_RESOURCES;
+  }
+  return RetErr;
+}
+
+UR_APIEXPORT ur_result_t UR_APICALL
+urContextGetInfo(ur_context_handle_t hContext, ur_context_info_t propName,
+                 size_t propSize, void *pPropValue, size_t *pPropSizeRet) {
+
+  UrReturnHelper ReturnValue(propSize, pPropValue, pPropSizeRet);
+
+  switch (uint32_t{propName}) {
+  case UR_CONTEXT_INFO_NUM_DEVICES:
+    return ReturnValue(1);
+  case UR_CONTEXT_INFO_DEVICES:
+    return ReturnValue(hContext->getDevice());
+  case UR_CONTEXT_INFO_REFERENCE_COUNT:
+    return ReturnValue(hContext->getReferenceCount());
+  case UR_CONTEXT_INFO_ATOMIC_MEMORY_ORDER_CAPABILITIES:
+  case UR_CONTEXT_INFO_ATOMIC_MEMORY_SCOPE_CAPABILITIES:
+  case UR_CONTEXT_INFO_ATOMIC_FENCE_ORDER_CAPABILITIES:
+  case UR_CONTEXT_INFO_ATOMIC_FENCE_SCOPE_CAPABILITIES: {
+    // These queries should be dealt with in context_impl.cpp by calling the
+    // queries of each device separately and building the intersection set.
+    setErrorMessage("These queries should have never come here.",
```

Comment:
Once you merge this the error message set here will be lost due to: https://github.com/intel/llvm/issues/10066 
I suggest switching to UR_CHECK_ERROR usage temporarily so that the error message is preserved: like this change: https://github.com/intel/llvm/pull/8303/commits/37aecb11e768fdb4ad3c590d83b3ac4f5e03db7b

Reply:
Actually, I think this should be `UR_RESULT_ERROR_INVALID_ENUMERATION` rather than `UR_RESULT_ERROR_ADAPTER_SPECIFIC` as the error comes from that this queries shouldn't be dealt with in UR (as the comment explains). So, I think it is more of an error from UR layer than CUDA layer, does that sound reasonable?

Reply:
or also could make sense to be `ur::assertion` if we should not return gracefully in this case.

Reply:
I have taken the path of making this return `UR_RESULT_ERROR_INVALID_ENUMERATION` as this should be a UR layer error, and the comment was explaining what happens enough to not support that.